airbyte-cdk/python/airbyte_cdk/test/utils/assertions.py

```diff
@@ -0,0 +1,28 @@
+# Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+
+import re
+
+from airbyte_cdk.test.entrypoint_wrapper import EntrypointOutput
+
+
+def is_in_logs(pattern: str, output: EntrypointOutput) -> bool:
+    """Check if any log message case-insensitive matches the pattern."""
+    return any(re.search(pattern, entry.log.message, flags=re.IGNORECASE) for entry in output.logs)
+
+
+def is_not_in_logs(pattern: str, output: EntrypointOutput) -> bool:
+    """Check if no log message matches the pattern."""
+    return not is_in_logs(pattern, output)
+
+
+def assert_good_read(output: EntrypointOutput, expected_record_count: int) -> None:
+    """Check if the output is successful read with an expected record count and no errors."""
+    assert len(output.errors) == 0
+    assert len(output.records) == expected_record_count
+    assert is_not_in_logs("error|exception", output)
+
+
+def assert_bad_read(output: EntrypointOutput, expected_record_count: int) -> None:
+    """Check if the output is unsuccessful read with an expected record count and errors."""
+    assert len(output.records) == expected_record_count
+    assert is_in_logs("error|exception", output)
```

Comment:
I feel this is very dangerous as there could be a log for an error that doesn't imply it is a "bad read". The same comment applies to `assert_good_read` as not all the the exceptions/errors have those in the name (for example, `BulkJobCanceled` in shopify). Would it be more accurate to rely on trace messages of a specific type?

Reply:
Yes, I think it's better to use message type. I'll change that. Regarding the error not implying a bad read, I wanted to provide a stricter assertion. I think a good read shouldn't have any exceptions, they may not be fatal, but doesn't smell good to have exceptions. These are just tools, if a test fails using this assertion, you are free to write a more relaxed set of assertions directly in your test, or create another common assertion that is useful.

Reply:
> I think a good read shouldn't have any exceptions

I'm not sure we can rely on that. We can debate if [this](https://github.com/airbytehq/airbyte/blob/148dccda4f0fc5fbec7e3d04805ee76eb1a413ca/airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py#L625) is considered failed or [this](https://github.com/airbytehq/airbyte/blob/148dccda4f0fc5fbec7e3d04805ee76eb1a413ca/airbyte-integrations/connectors/source-hubspot/source_hubspot/streams.py#L229-L231) should be considered `bad` or `good` but they are only flagged as a warning but they have `error` in the logs. I wouldn't be surprised that debug logs eventually add more messages that could contain those keywords too

Reply:
Removed the controversial "good vs bad" and used normal assertions.